File: Arrays/ThreeSum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    const length = nums.length;

    if(length < 3) {
        return [];
    }

    const sorted = nums.sort(function(a, b) {return a - b});
    const seen = [];
    const ans = [];
    
    for(let i  = 0; i < nums.length - 2; i += 1) {

        let currNum = sorted[i];
    
        if(i > 0 && sorted[i] == sorted[i - 1]) {
            continue;
        }

        let leftPointer = i + 1;
        let rightPointer = length - 1;
        let target = -1 * currNum;
        
        while(leftPointer != rightPointer) {
           let sum = sorted[leftPointer] + sorted[rightPointer];
           if(sum > target) {
                rightPointer = rightPointer - 1;
            }

            else if (sum < target) {
                leftPointer = leftPointer + 1;
            }

            else {
                
                ans.push([currNum, sorted[leftPointer], sorted[rightPointer]]);

                while(sorted[leftPointer] == sorted[leftPointer + 1]) {
                    leftPointer += 1;
                }

                
                while(sorted[rightPointer] == sorted[rightPointer - 1]) {
                    rightPointer -= 1;
                }

                leftPointer += 1;
                rightPointer -= 1;
            }

        }

        seen.push(currNum);

    }

    return ans;
}


console.log(threeSum([-1,0,1,2,-1,-4]));

File: Arrays/ThreeSum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        length = len(nums)
        seen = {}
        index = 0
        solution = []

        if length < 3:
            return []

        nums.sort()

        while index < length:

            if seen.get(nums[index]) != None :
                index += 1
                continue

            
            right = length - 1
            left = index + 1
            target = -1 * nums[index]
            seen[nums[index]] = index
            while left < right:

                left_num = nums[left]
                right_num = nums[right]
                sum = left_num + right_num

                if sum < target:
                    left += 1
                elif sum > target:
                    right -= 1
                else:
                    solution.append([nums[index], nums[left], nums[right]])
                    left += 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
                

            
            index += 1
        
        return solution

            


